src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NodeFactory.cs

```diff
@@ -341,6 +341,11 @@ private void CreateNodeCaches()
 
         public IEETypeNode NecessaryTypeSymbol(TypeDesc type)
         {
+            if (!_compilationModuleGroup.IsSingleFileCompilation && EETypeNode.IsTypeNodeShareable(type))
```

Comment:
Yeah I can do that. The problem is `ShouldProduceFullType` returns true for all types in shared library mode so we end up fully building generic type defs and maybe other stuff we don't want to be constructed. I can probably be more choosy about what gets a full type in shared libraries. I'll have a play and report back.